File: tools/tracer/src/tracer/commands/blame.py
"""`trace blame` — scoped, symbol-aware blame.

Returns blame information for a file, scoped to either the whole file,
an explicit line range (`--lines L1:L2`), or a named symbol resolved to
its line range via lizard. Consecutive lines sharing the same commit
are collapsed into one region. Each region carries the commit subject
so the caller never needs a follow-up `git show` to learn the "why".
"""

from __future__ import annotations

import json
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

import click
import lizard

from tracer import cache, file_facts
from tracer.deps import require_dependencies


def _parse_lines(spec: str) -> tuple[int, int]:
    """Parse 'L1:L2' into (start, end). Inclusive on both ends."""
    if ":" not in spec:
        raise click.BadParameter("--lines must be in form L1:L2")
    raw_start, raw_end = spec.split(":", 1)
    try:
        start = int(raw_start)
        end = int(raw_end)
    except ValueError as exc:
        raise click.BadParameter("--lines must be integers, e.g. 10:42") from exc
    if start < 1 or end < start:
        raise click.BadParameter("--lines start must be >= 1 and <= end")
    return start, end


def _resolve_symbol_range(file: Path, symbol: str) -> tuple[int, int] | None:
    """Resolve a symbol name to its (start_line, end_line) range via lizard.

    Matches either an exact function name or a qualified suffix
    (e.g. `Class.method`) — same matcher shape as `trace read`.
    """
    try:
        parsed = lizard.analyze_file(str(file))
    except Exception:
        return None
    target = next(
        (
            function
            for function in parsed.function_list
            if function.name == symbol or function.name.endswith(f".{symbol}")
        ),
        None,
    )
    if target is None:
        return None
    return target.start_line, target.end_line


UNTRACKED_MARKERS = ("no such path", "is outside repository", "not in a git repository")


def _run_blame(file: Path, line_range: tuple[int, int] | None) -> str:
    """Run `git blame --line-porcelain`, optionally scoped to a range.

    For untracked files git blame fails with "no such path … in HEAD";
    we synthesize a porcelain stream that marks every line uncommitted
    so the command returns valid output instead of crashing.
    """
    command = ["git", "blame", "--line-porcelain"]
    if line_range is not None:
        start, end = line_range
        command += ["-L", f"{start},{end}"]
    command += ["--", str(file)]
    result = subprocess.run(
        command,
        cwd=file.parent,
        capture_output=True,
        text=True,
        check=False,
        timeout=30,
    )
    if result.returncode == 0:
        return result.stdout
    stderr_lower = result.stderr.lower()
    if any(marker in stderr_lower for marker in UNTRACKED_MARKERS):
        return _synthesize_untracked_porcelain(file, line_range)
    message = result.stderr.strip() or "git blame failed"
    click.echo(f"Error: {message}", err=True)
    sys.exit(1)


def _synthesize_untracked_porcelain(
    file: Path, line_range: tuple[int, int] | None
) -> str:
    """Build porcelain output for an untracked file: every line uncommitted."""
    with open(file, encoding="utf-8", errors="replace") as handle:
        source_lines = handle.read().splitlines()
    if line_range is None:
        start, end = 1, len(source_lines)
    else:
        start, end = line_range
        end = min(end, len(source_lines))
    zero_sha = "0" * 40
    now = int(datetime.now(tz=timezone.utc).timestamp())
    chunks: list[str] = []
    for line_number in range(start, end + 1):
        content = source_lines[line_number - 1] if line_number <= len(source_lines) else ""
        chunks.append(
            f"{zero_sha} {line_number} {line_number} 1\n"
            f"author Not Committed Yet\n"
            f"author-time {now}\n"
            f"author-tz +0000\n"
            f"summary (uncommitted)\n"
            f"\t{content}\n"
        )
    return "".join(chunks)


def _parse_porcelain(output: str) -> list[dict]:
    """Parse `git blame --line-porcelain` output into per-line records.

    Porcelain groups: each line begins with a header
    `<sha> <orig-line> <final-line> [<group-size>]`, followed by
    key/value attribute lines, then a tab-prefixed content line.
    The full attribute block only appears the first time a commit is
    seen; subsequent lines carry only the header + content. We track
    commit attributes in a dict keyed by sha to fill in repeats.
    """
    commits: dict[str, dict] = {}
    lines: list[dict] = []
    current_sha: str | None = None
    current_line: int | None = None
    pending: dict[str, str] = {}

    for raw in output.splitlines():
        if raw.startswith("\t"):
            # Content line — closes the current per-line record.
            if current_sha is None or current_line is None:
                continue
            attributes = commits.setdefault(current_sha, {})
            attributes.update({k: v for k, v in pending.items() if k not in attributes})
            lines.append(
                {
                    "line": current_line,
                    "sha": current_sha,
                    "author": attributes.get("author", "unknown"),
                    "author_time": int(attributes.get("author-time", "0")),
                    "author_tz": attributes.get("author-tz", "+0000"),
                    "summary": attributes.get("summary", ""),
                }
            )
            current_sha = None
            current_line = None
            pending = {}
            continue

        parts = raw.split(" ", 3)
        if current_sha is None and len(parts) >= 3 and len(parts[0]) >= 7:
            # Header line: <sha> <orig> <final> [<group-size>]
            current_sha = parts[0]
            try:
                current_line = int(parts[2])
            except ValueError:
                current_sha = None
                current_line = None
            continue

        # Attribute line within the current group.
        if " " in raw:
            key, _, value = raw.partition(" ")
            pending[key] = value

    return lines


def _collapse_regions(lines: list[dict]) -> list[dict]:
    """Collapse consecutive lines sharing the same commit into regions."""
    regions: list[dict] = []
    for entry in lines:
        if (
            regions
            and regions[-1]["sha"] == entry["sha"]
            and regions[-1]["line_end"] + 1 == entry["line"]
        ):
            regions[-1]["line_end"] = entry["line"]
            continue
        regions.append(
            {
                "line_start": entry["line"],
                "line_end": entry["line"],
                "sha": entry["sha"],
                "author": entry["author"],
                "author_time": entry["author_time"],
                "author_tz": entry["author_tz"],
                "subject": entry["summary"],
            }
        )
    return regions


def _is_uncommitted(sha: str) -> bool:
    return sha.startswith("0" * 8)


def _short_sha(sha: str) -> str:
    return "uncommitted" if _is_uncommitted(sha) else sha[:8]


def _subject(region: dict) -> str:
    """Subject of the region's commit; replaces git's verbose placeholder for uncommitted lines."""
    if _is_uncommitted(region["sha"]):
        return "(uncommitted change)"
    return region["subject"] or "(no subject)"


def _isoformat_date(author_time: int, author_tz: str) -> str:
    """Render author_time as YYYY-MM-DD in the author's timezone."""
    if author_time == 0:
        return "—"
    moment = datetime.fromtimestamp(author_time, tz=timezone.utc)
    return moment.strftime("%Y-%m-%d")


def _humanize_age(author_time: int) -> str:
    if author_time == 0:
        return "uncommitted"
    delta = datetime.now(tz=timezone.utc) - datetime.fromtimestamp(
        author_time, tz=timezone.utc
    )
    seconds = int(delta.total_seconds())
    if seconds < 60:
        return f"{seconds}s ago"
    minutes = seconds // 60
    if minutes < 60:
        return f"{minutes}m ago"
    hours = minutes // 60
    if hours < 24:
        return f"{hours}h ago"
    days = hours // 24
    if days < 30:
        return f"{days}d ago"
    months = days // 30
    if months < 12:
        return f"{months}mo ago"
    years = days // 365
    return f"{years}y ago"


def _format_region(region: dict) -> str:
    line_start = region["line_start"]
    line_end = region["line_end"]
    span = f"L{line_start}" if line_start == line_end else f"L{line_start}-L{line_end}"
    sha = _short_sha(region["sha"])
    date = _isoformat_date(region["author_time"], region["author_tz"])
    age = _humanize_age(region["author_time"])
    subject = _subject(region)
    author = region["author"]
    return f"  {span:<14} {sha:<11} {date}  {age:<10}  {author:<22} {subject}"


@click.command()
@click.argument("file", type=click.Path(exists=True, dir_okay=False))
@click.argument("symbol", required=False)
@click.option(
    "--lines",
    "lines_spec",
    metavar="L1:L2",
    help="Blame only this inclusive line range; mutually exclusive with SYMBOL.",
)
@click.option("--json", "as_json", is_flag=True, help="Machine-parseable JSON output.")
def command(file: str, symbol: str | None, lines_spec: str | None, as_json: bool) -> None:
    """Blame a file, line range, or symbol — collapsed into regions.

    Three scoping modes:

    \b
      trace blame <file>                  whole file
      trace blame <file> <symbol>         just the symbol's line range
      trace blame <file> --lines L1:L2    just that explicit line range
    """
    require_dependencies()

    if symbol is not None and lines_spec is not None:
        click.echo("Error: SYMBOL and --lines are mutually exclusive", err=True)
        sys.exit(2)

    path = Path(file).resolve()

    line_range: tuple[int, int] | None = None
    scope = "file"
    if symbol is not None:
        resolved = _resolve_symbol_range(path, symbol)
        if resolved is None:
            click.echo(f"Error: symbol '{symbol}' not found in {file}", err=True)
            sys.exit(2)
        line_range = resolved
        scope = "symbol"
    elif lines_spec is not None:
        line_range = _parse_lines(lines_spec)
        scope = "lines"

    porcelain = _run_blame(path, line_range)
    blame_lines = _parse_porcelain(porcelain)
    regions = _collapse_regions(blame_lines)

    facts = file_facts.get(path)
    repo_root = cache.repo_root_for(path)
    try:
        display_file = str(path.relative_to(repo_root.resolve()))
    except ValueError:
        display_file = str(path)

    if as_json:
        click.echo(
            json.dumps(
                {
                    "file": display_file,
                    "language": facts.language if facts else None,
                    "scope": scope,
                    "symbol": symbol,
                    "line_range": (
                        {"start": line_range[0], "end": line_range[1]}
                        if line_range
                        else None
                    ),
                    "regions": [
                        {
                            "line_start": region["line_start"],
                            "line_end": region["line_end"],
                            "sha": _short_sha(region["sha"]),
                            "author": region["author"],
                            "date": _isoformat_date(
                                region["author_time"], region["author_tz"]
                            ),
                            "age": _humanize_age(region["author_time"]),
                            "subject": _subject(region),
                        }
                        for region in regions
                    ],
                    "region_count": len(regions),
                    "line_count": len(blame_lines),
                },
                indent=2,
            )
        )
        return

    header = f"# {display_file}"
    if scope == "symbol":
        header += f" :: {symbol}  (L{line_range[0]}-L{line_range[1]})"
    elif scope == "lines":
        header += f"  L{line_range[0]}-L{line_range[1]}"
    click.echo(header)
    click.echo(
        f"Regions: {len(regions)}  Lines blamed: {len(blame_lines)}"
    )
    click.echo("")
    click.echo(
        f"  {'lines':<14} {'commit':<11} {'date':<10}  {'age':<10}  "
        f"{'author':<22} subject"
    )
    for region in regions:
        click.echo(_format_region(region))

File: tools/tracer/src/commands/blame.rs
//! `trace blame` — scoped, symbol-aware blame.
//! Symbol-range resolution is AST-derived via `crate::ccn` (maps a symbol
//! name to its line span). Porcelain parsing, region collapse, and
//! untracked-file synthesis emit per-region commit + subject so no
//! follow-up `git show` is needed.

use crate::{cache, ccn, file_facts};
use anyhow::{bail, Result};
use serde_json::{json, Value};
use std::path::Path;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

fn parse_lines(spec: &str) -> Result<(i64, i64)> {
    let (a, b) = spec
        .split_once(':')
        .ok_or_else(|| anyhow::anyhow!("--lines must be in form L1:L2"))?;
    let start: i64 = a
        .parse()
        .map_err(|_| anyhow::anyhow!("--lines must be integers, e.g. 10:42"))?;
    let end: i64 = b
        .parse()
        .map_err(|_| anyhow::anyhow!("--lines must be integers, e.g. 10:42"))?;
    if start < 1 || end < start {
        bail!("--lines start must be >= 1 and <= end");
    }
    Ok((start, end))
}

/// Resolve a symbol name to its (start_line, end_line) range via the AST CCN
/// backend. Matches an exact function name or a qualified suffix
/// (`Class.method`) — the same matcher shape `trace read` uses. end_line
/// is start_line + nloc - 1 (AST line span).
fn resolve_symbol_range(file: &Path, symbol: &str) -> Option<(i64, i64)> {
    let source = std::fs::read(file).ok()?;
    let functions = ccn::analyze(&source, &file.to_string_lossy())?;
    let suffix = format!(".{symbol}");
    let target = functions
        .iter()
        .find(|f| f.name == symbol || f.name.ends_with(&suffix))?;
    Some((target.start_line, target.start_line + target.nloc - 1))
}

const UNTRACKED_MARKERS: &[&str] = &[
    "no such path",
    "is outside repository",
    "not in a git repository",
];

fn run_blame(file: &Path, line_range: Option<(i64, i64)>) -> Result<String> {
    let mut args: Vec<String> = vec!["blame".into(), "--line-porcelain".into()];
    if let Some((s, e)) = line_range {
        args.push("-L".into());
        args.push(format!("{s},{e}"));
    }
    args.push("--".into());
    args.push(file.to_string_lossy().to_string());
    let parent = file.parent().unwrap_or_else(|| Path::new("."));
    let out = Command::new("git")
        .args(&args)
        .current_dir(parent)
        .output()?;
    if out.status.success() {
        return Ok(String::from_utf8_lossy(&out.stdout).to_string());
    }
    let stderr = String::from_utf8_lossy(&out.stderr).to_lowercase();
    if UNTRACKED_MARKERS.iter().any(|m| stderr.contains(m)) {
        return Ok(synthesize_untracked_porcelain(file, line_range));
    }
    let msg = String::from_utf8_lossy(&out.stderr).trim().to_string();
    bail!(if msg.is_empty() {
        "git blame failed".to_string()
    } else {
        msg
    });
}

fn synthesize_untracked_porcelain(file: &Path, line_range: Option<(i64, i64)>) -> String {
    let text = std::fs::read_to_string(file).unwrap_or_default();
    let source_lines: Vec<&str> = text.split('\n').collect();
    // splitlines() drops a trailing newline's empty segment; emulate that.
    let line_count = if text.ends_with('\n') && !source_lines.is_empty() {
        source_lines.len() - 1
    } else {
        source_lines.len()
    };
    let (start, end) = match line_range {
        None => (1i64, line_count as i64),
        Some((s, e)) => (s, e.min(line_count as i64)),
    };
    let zero_sha = "0".repeat(40);
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    let mut chunks = String::new();
    for ln in start..=end {
        let content = if ln >= 1 && (ln as usize) <= line_count {
            source_lines[(ln - 1) as usize]
        } else {
            ""
        };
        chunks.push_str(&format!(
            "{zero_sha} {ln} {ln} 1\n\
             author Not Committed Yet\n\
             author-time {now}\n\
             author-tz +0000\n\
             summary (uncommitted)\n\
             \t{content}\n"
        ));
    }
    chunks
}

#[derive(Clone)]
struct BlameLine {
    line: i64,
    sha: String,
    author: String,
    author_time: i64,
    author_tz: String,
    summary: String,
}

/// Parse `git blame --porcelain`. The commit attribute block appears once
/// per sha; later lines carry only header + content, so attributes are
/// memoized.
fn parse_porcelain(output: &str) -> Vec<BlameLine> {
    use std::collections::HashMap;
    let mut commits: HashMap<String, HashMap<String, String>> = HashMap::new();
    let mut lines: Vec<BlameLine> = Vec::new();
    let mut current_sha: Option<String> = None;
    let mut current_line: Option<i64> = None;
    let mut pending: HashMap<String, String> = HashMap::new();

    for raw in output.split('\n') {
        if let Some(content) = raw.strip_prefix('\t') {
            let (sha, line) = match (&current_sha, current_line) {
                (Some(s), Some(l)) => (s.clone(), l),
                _ => continue,
            };
            let attrs = commits.entry(sha.clone()).or_default();
            for (k, v) in &pending {
                attrs.entry(k.clone()).or_insert_with(|| v.clone());
            }
            let _ = content;
            lines.push(BlameLine {
                line,
                sha: sha.clone(),
                author: attrs.get("author").cloned().unwrap_or_else(|| "unknown".into()),
                author_time: attrs
                    .get("author-time")
                    .and_then(|s| s.parse().ok())
                    .unwrap_or(0),
                author_tz: attrs
                    .get("author-tz")
                    .cloned()
                    .unwrap_or_else(|| "+0000".into()),
                summary: attrs.get("summary").cloned().unwrap_or_default(),
            });
            current_sha = None;
            current_line = None;
            pending = HashMap::new();
            continue;
        }
        let parts: Vec<&str> = raw.splitn(4, ' ').collect();
        if current_sha.is_none() && parts.len() >= 3 && parts[0].len() >= 7 {
            current_sha = Some(parts[0].to_string());
            match parts[2].parse::<i64>() {
                Ok(n) => current_line = Some(n),
                Err(_) => {
                    current_sha = None;
                    current_line = None;
                }
            }
            continue;
        }
        if let Some((k, v)) = raw.split_once(' ') {
            pending.insert(k.to_string(), v.to_string());
        }
    }
    lines
}

struct Region {
    line_start: i64,
    line_end: i64,
    sha: String,
    author: String,
    author_time: i64,
    author_tz: String,
    subject: String,
}

fn collapse_regions(lines: &[BlameLine]) -> Vec<Region> {
    let mut regions: Vec<Region> = Vec::new();
    for entry in lines {
        if let Some(last) = regions.last_mut() {
            if last.sha == entry.sha && last.line_end + 1 == entry.line {
                last.line_end = entry.line;
                continue;
            }
        }
        regions.push(Region {
            line_start: entry.line,
            line_end: entry.line,
            sha: entry.sha.clone(),
            author: entry.author.clone(),
            author_time: entry.author_time,
            author_tz: entry.author_tz.clone(),
            subject: entry.summary.clone(),
        });
    }
    regions
}

fn is_uncommitted(sha: &str) -> bool {
    sha.starts_with("00000000")
}

fn short_sha(sha: &str) -> String {
    if is_uncommitted(sha) {
        "uncommitted".into()
    } else {
        sha.chars().take(8).collect()
    }
}

fn subject(region: &Region) -> String {
    if is_uncommitted(&region.sha) {
        return "(uncommitted change)".into();
    }
    if region.subject.is_empty() {
        "(no subject)".into()
    } else {
        region.subject.clone()
    }
}

/// YYYY-MM-DD in UTC from author_time; the tz is informational only.
fn isoformat_date(author_time: i64, _author_tz: &str) -> String {
    if author_time == 0 {
        return "—".into();
    }
    ymd_from_unix(author_time)
}

fn humanize_age(author_time: i64) -> String {
    if author_time == 0 {
        return "uncommitted".into();
    }
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    let seconds = now - author_time;
    if seconds < 60 {
        return format!("{seconds}s ago");
    }
    let minutes = seconds / 60;
    if minutes < 60 {
        return format!("{minutes}m ago");
    }
    let hours = minutes / 60;
    if hours < 24 {
        return format!("{hours}h ago");
    }
    let days = hours / 24;
    if days < 30 {
        return format!("{days}d ago");
    }
    let months = days / 30;
    if months < 12 {
        return format!("{months}mo ago");
    }
    let years = days / 365;
    format!("{years}y ago")
}

/// UTC date YYYY-MM-DD from a unix timestamp (civil calc, same as
/// git_activity::unix_to_ymd which is module-private).
fn ymd_from_unix(secs: i64) -> String {
    let days = secs.div_euclid(86400);
    let z = days + 719468;
    let era = if z >= 0 { z } else { z - 146096 } / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    format!("{:04}-{:02}-{:02}", y, m, d)
}

fn format_region(region: &Region) -> String {
    let span = if region.line_start == region.line_end {
        format!("L{}", region.line_start)
    } else {
        format!("L{}-L{}", region.line_start, region.line_end)
    };
    format!(
        "  {:<14} {:<11} {}  {:<10}  {:<22} {}",
        span,
        short_sha(&region.sha),
        isoformat_date(region.author_time, &region.author_tz),
        humanize_age(region.author_time),
        region.author,
        subject(region),
    )
}

pub fn run(
    file: &Path,
    symbol: Option<&str>,
    lines_spec: Option<&str>,
    as_json: bool,
) -> Result<Value> {
    if symbol.is_some() && lines_spec.is_some() {
        bail!("SYMBOL and --lines are mutually exclusive");
    }
    if !file.is_file() {
        bail!("file not found: {}", file.display());
    }
    let path = file.canonicalize().unwrap_or_else(|_| cache::absolutize(file));

    let mut line_range: Option<(i64, i64)> = None;
    let mut scope = "file";
    if let Some(sym) = symbol {
        match resolve_symbol_range(&path, sym) {
            Some(r) => {
                line_range = Some(r);
                scope = "symbol";
            }
            None => bail!("symbol '{}' not found in {}", sym, file.display()),
        }
    } else if let Some(spec) = lines_spec {
        line_range = Some(parse_lines(spec)?);
        scope = "lines";
    }

    let porcelain = run_blame(&path, line_range)?;
    let blame_lines = parse_porcelain(&porcelain);
    let regions = collapse_regions(&blame_lines);

    let repo_root = cache::worktree_root_for(&path).unwrap_or_else(|| cache::display_root(&path));
    let facts = file_facts::get(&path, &repo_root, None);
    let display_file = cache::relative_to_root(&path, &repo_root);

    let payload = json!({
        "file": display_file,
        "language": facts.as_ref().and_then(|f| f.language.clone()),
        "scope": scope,
        "symbol": symbol,
        "line_range": line_range.map(|(s, e)| json!({"start": s, "end": e})),
        "regions": regions.iter().map(|r| json!({
            "line_start": r.line_start,
            "line_end": r.line_end,
            "sha": short_sha(&r.sha),
            "author": r.author,
            "date": isoformat_date(r.author_time, &r.author_tz),
            "age": humanize_age(r.author_time),
            "subject": subject(r),
        })).collect::<Vec<_>>(),
        "region_count": regions.len(),
        "line_count": blame_lines.len(),
    });

    if as_json {
        return Ok(payload);
    }

    let mut header = format!("# {display_file}");
    if scope == "symbol" {
        let (s, e) = line_range.unwrap();
        header += &format!(" :: {}  (L{}-L{})", symbol.unwrap(), s, e);
    } else if scope == "lines" {
        let (s, e) = line_range.unwrap();
        header += &format!("  L{s}-L{e}");
    }
    println!("{header}");
    println!(
        "Regions: {}  Lines blamed: {}",
        regions.len(),
        blame_lines.len()
    );
    println!();
    println!(
        "  {:<14} {:<11} {:<10}  {:<10}  {:<22} subject",
        "lines", "commit", "date", "age", "author"
    );
    for region in &regions {
        println!("{}", format_region(region));
    }
    Ok(payload)
}
